### PR TITLE
🌱 Add sprig functions to ClusterClass templates

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds sprig functions to ClusterClass templates

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/6091

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -45,7 +45,18 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
-require golang.org/x/net v0.0.0-20210825183410-e898025ed96a
+require (
```
Comment (sbueringer) [inline go.mod:48]: Can you resolve those blocks by adding the dependencies to the existing ones + running `make modules`. I assume that's like with goimports, that only the blocks are formatted internally.

**Review by sbueringer — COMMENTED** [review]
Looks good, very nice small change :)

I think you got all places. Essentially we have two "template" variables and for each of those we have a validation and a place where it's used during patching.

**Review by killianmuldoon — COMMENTED** [review]
lgtm with the go.mod update. 

I think we should open an issue for more comprehensive patching documentation as this gets more complex. A few examples of valid patches and the functions we're offering here (doesn't have to be exhaustive.) I'll take that as an issue if you'd like.

**Comment by fabriziopandini** [comment]
> I think we should open an issue for more comprehensive patching documentation as this gets more complex. A few examples of valid patches and the functions we're offering here (doesn't have to be exhaustive.) I'll take that as an issue if you'd like.

Some examples would be nice, but I would wait for providers to start using this so we can provide real use cases.

**Comment by sbueringer** [comment]
Thx!
/lgtm

**Review by killianmuldoon — APPROVED** [review]
/lgtm

File: docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md [hunk 2]
```diff
@@ -587,6 +587,9 @@ constant default value which can be specified in the schema is not enough.
           # If .vnetName is set, it is used. Otherwise, we will use `{{.builtin.cluster.name}}-vnet`.  
           template: "{{ if .vnetName }}{{.vnetName}}{{else}}{{.builtin.cluster.name}}-vnet{{end}}"
 ```
+When writing templates, a subset of functions from [sprig](http://masterminds.github.io/sprig/) can be used to
```
Comment (vincepri) [inline docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:590]:
```suggestion
When writing templates, a subset of functions from [the sprig library](http://masterminds.github.io/sprig/) can be used to
```

File: docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md [hunk 3]
```diff
@@ -587,6 +587,9 @@ constant default value which can be specified in the schema is not enough.
...
+write expressing like e.g. `{{ .name | upper }}`. Only functions that are guaranteed to evaluate to the same result
+for a given input are allowed (e.g. `upper` or `max` can be used, while `now` or `randAlpha` can not be used).
```
Comment (vincepri) [inline docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:593]: Can we document all the available functions in here?

Comment (sbueringer) [reply docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:593]: Can we instead refer to e.g. that documentation page: http://masterminds.github.io/sprig/

I would really prefer not duplicating part of the docs (even if it's only a list of funcs).

File: docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md [hunk 4]
```diff
@@ -587,6 +587,9 @@ constant default value which can be specified in the schema is not enough.
           # If .vnetName is set, it is used. Otherwise, we will use `{{.builtin.cluster.name}}-vnet`.  
           template: "{{ if .vnetName }}{{.vnetName}}{{else}}{{.builtin.cluster.name}}-vnet{{end}}"
 ```
+When writing templates, a subset of functions from [sprig](http://masterminds.github.io/sprig/) can be used to
+write expressing like e.g. `{{ .name | upper }}`. Only functions that are guaranteed to evaluate to the same result
```
Comment (ykakarap) [inline docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:591]:
```suggestion
write expressions like e.g. `{{ .name | upper }}`. Only functions that are guaranteed to evaluate to the same result
```

File: go.mod [hunk 5]
```diff
@@ -4,6 +4,7 @@ go 1.17
 
 require (
 	github.com/MakeNowJust/heredoc v1.0.0
+	github.com/Masterminds/sprig v2.22.0+incompatible
```
Comment (ykakarap) [inline go.mod:7]: Any specific reason we didnt pick up the newer version of spirg library `v3.2.0`?
`v2.22.0` is more than 2 years old.

The v3 release seems to be having some improvements related to edge cases to their semver functions.

File: docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md (hunk 3 above)
Comment (fabriziopandini) [reply docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:593]: there are over 70 template functions in sprig, I don't this it is doable to document here.

Comment (vincepri) [reply docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:593]: Linking or importing works, from a user experience having them in one place (even if generate) is much more useful when implementing a clusterclass

File: go.mod (hunk 5 above)
Comment (sbueringer) [reply go.mod]: @ykakarap nice catch!!

**Review by ykakarap — COMMENTED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
As per https://github.com/kubernetes-sigs/cluster-api/issues/6091#issuecomment-1036432595 
/cherry-pick release-1.1

**Comment by sbueringer** [comment]
Manual cherry-pick: https://github.com/kubernetes-sigs/cluster-api/pull/6191

**Pull request merged**